pkg/env/var.go

```diff
@@ -0,0 +1,243 @@
+	Hidden bool
+
+	// Mark this variable as deprecated when generating usage information.
+	Deprecated bool
+
+	// The type of the variable's value
+	Type VarType
+}
+
+// StringVar represents a single string environment variable.
+type StringVar struct {
+	Var
+}
+
+// BoolVar represents a single boolean environment variable.
+type BoolVar struct {
+	Var
+}
+
+// IntVar represents a single integer environment variable.
+type IntVar struct {
+	Var
+}
+
+// FloatVar represents a single floating-point environment variable.
+type FloatVar struct {
+	Var
+}
+
+// DurationVar represents a single duration environment variable.
+type DurationVar struct {
+	Var
+}
+
+var allVars = make(map[string]Var)
+var mutex sync.Mutex
+
+// Returns a description of this process' environment variables, sorted by name.
+func VarDescriptions() []Var {
+	mutex.Lock()
+	sorted := make([]Var, 0, len(allVars))
+	for _, v := range allVars {
+		sorted = append(sorted, v)
+	}
+	mutex.Unlock()
+
+	sort.Slice(sorted, func(i, j int) bool {
+		return sorted[i].Name < sorted[j].Name
+	})
+
+	return sorted
+}
+
+// RegisterStringVar registers a new string environment variable.
+func RegisterStringVar(name string, defaultValue string, description string) StringVar {
+	v := Var{Name: name, DefaultValue: defaultValue, Description: description, Type: STRING}
+	RegisterVar(v)
+	return StringVar{v}
+}
+
+// RegisterBoolVar registers a new boolean environment variable.
+func RegisterBoolVar(name string, defaultValue bool, description string) BoolVar {
+	v := Var{Name: name, DefaultValue: strconv.FormatBool(defaultValue), Description: description, Type: BOOL}
+	RegisterVar(v)
+	return BoolVar{v}
+}
+
+// RegisterIntVar registers a new integer environment variable.
+func RegisterIntVar(name string, defaultValue int, description string) IntVar {
+	v := Var{Name: name, DefaultValue: strconv.FormatInt(int64(defaultValue), 10), Description: description, Type: INT}
+	RegisterVar(v)
+	return IntVar{v}
+}
+
+// RegisterFloatVar registers a new floating-point environment variable.
+func RegisterFloatVar(name string, defaultValue float64, description string) FloatVar {
+	v := Var{Name: name, DefaultValue: strconv.FormatFloat(defaultValue, 'G', -1, 64), Description: description, Type: FLOAT}
+	RegisterVar(v)
+	return FloatVar{v}
+}
+
+// RegisterDurationVar registers a new duration environment variable.
+func RegisterDurationVar(name string, defaultValue time.Duration, description string) DurationVar {
+	v := Var{Name: name, DefaultValue: defaultValue.String(), Description: description, Type: DURATION}
+	RegisterVar(v)
+	return DurationVar{v}
+}
+
+// RegisterVar registers a generic environment variable.
+func RegisterVar(v Var) {
+	mutex.Lock()
+
+	if old, ok := allVars[v.Name]; !ok || old.Description == "" {
+		allVars[v.Name] = v // last one with a description wins if there are multiple enlistments of the same variable in a single process
```

Comment:
should you log if a variable with the same name is registered more than once, but with different values?